### PR TITLE
JAMES-3313 Introduce a generic "DeDuplicatingBlobStore"

### PR DESCRIPTION
**Pull request by chibenwa**
- Drops the HybridBlobStore
 - Creates a DeDuplicatingBlobStore
 - Cassandra & Memory relies on it

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[cfee07b3486db7f963a1383caac7a26c7bb0cc97] [ERROR] createShouldSetAMailboxIdForMailbox  Time elapsed: 0.018 s  <<< ERROR!
[cfee07b3486db7f963a1383caac7a26c7bb0cc97] com.google.inject.CreationException: 
[cfee07b3486db7f963a1383caac7a26c7bb0cc97] Unable to create injector, see the following errors:
[cfee07b3486db7f963a1383caac7a26c7bb0cc97] 
[cfee07b3486db7f963a1383caac7a26c7bb0cc97] 1) An exception was caught and reported. Message: unconfigured table blobsinbucket
[cfee07b3486db7f963a1383caac7a26c7bb0cc97]   at com.google.inject.internal.InjectorShell$Builder.build(InjectorShell.java:137) (via modules: com.google.inject.util.Modules$CombinedModule -> org.apache.james.mailbox.cassandra.mail.utils.GuiceUtils$$Lambda$552/0x000000010054e040)
[cfee07b3486db7f963a1383caac7a26c7bb0cc97] 
[cfee07b3486db7f963a1383caac7a26c7bb0cc97] 2) No implementation for org.apache.james.blob.api.BlobStore was bound.
[cfee07b3486db7f963a1383caac7a26c7bb0cc97]   at org.apache.james.mailbox.cassandra.mail.utils.GuiceUtils.lambda$commonModules$2(GuiceUtils.java:62) (via modules: com.google.inject.util.Modules$CombinedModule -> org.apache.james.mailbox.cassandra.mail.utils.GuiceUtils$$Lambda$552/0x000000010054e040)
```

**Comment by chibenwa** [comment]
```
[INFO] Apache James :: Server :: Cassandra - guice injection FAILURE [  02:00 h]

[7da75f93c74e92ea6c2891e0c157c3abdda3182c] [ERROR]   JamesCapabilitiesServerTest.startShouldFailWhenNoUniqueIDCapability:138 
[7da75f93c74e92ea6c2891e0c157c3abdda3182c] Expecting:
[7da75f93c74e92ea6c2891e0c157c3abdda3182c]   <com.google.inject.CreationException: Unable to create injector, see the following errors:
[7da75f93c74e92ea6c2891e0c157c3abdda3182c] 
[7da75f93c74e92ea6c2891e0c157c3abdda3182c] 1) No implementation for org.apache.james.blob.api.BucketName annotated with @com.google.inject.name.Named(value="cassandraDefault") was bound.
[7da75f93c74e92ea6c2891e0c157c3abdda3182c]   Did you mean?
[7da75f93c74e92ea6c2891e0c157c3abdda3182c]     org.apache.james.blob.api.BucketName annotated with @com.google.inject.name.Named(value="defaultBucket") bound  at org.apache.james.modules.mailbox.CassandraBlobStoreDependenciesModule.configure(CassandraBlobStoreDependenciesModule.java:42) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.mailbox.CassandraBlobStoreDependenciesModule)
[7da75f93c74e92ea6c2891e0c157c3abdda3182c] 
[7da75f93c74e92ea6c2891e0c157c3abdda3182c]   while locating org.apache.james.blob.api.BucketName annotated with @com.google.inject.name.Named(value="cassandraDefault")
[7da75f93c74e92ea6c2891e0c157c3abdda3182c]     for the 4th parameter of org.apache.james.blob.cassandra.CassandraDumbBlobStore.<init>(CassandraDumbBlobStore.java:64)
[7da75f93c74e92ea6c2891e0c157c3abdda3182c]   at org.apache.james.modules.mailbox.CassandraBlobStoreModule.configure(CassandraBlobStoreModule.java:35) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.mailbox.CassandraBlobStoreModule)
```

**Comment by chibenwa** [comment]
`[bc934002c6274c8f2c827f2d4ef039c1c5a2fa94] [ERROR] Failed to execute goal on project event-sourcing-pojo: Could not resolve dependencies for project org.apache.james:event-sourcing-pojo:jar:3.6.0-SNAPSHOT: Failed to collect dependencies at com.google.guava:guava:jar:25.1-jre -> org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Failed to read artifact descriptor for org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Could not transfer artifact org.codehaus.mojo:mojo-parent:pom:34 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/org/codehaus/mojo/mojo-parent/34/mojo-parent-34.pom.part (No such file or directory) -> [Help 1]`

test this please

**Comment by chibenwa** [comment]
`Caused by: java.io.IOException: remote file operation failed: /home/jenkins/build/workspace/1005/.8f4e8ce3 at hudson.remoting.Channel@2f0d650c:ci-james-07: hudson.remoting.ChannelClosedException: channel is already closed`

test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
\o/ It passed *mvn* tests

I will open a JIRA ticket, and better document that the hybrid blobStore had been removed.

File: server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  private def saveAndGenerateBlobId(bucketName: BucketName, hashingInputStream: HashingInputStream, fileBackedOutputStream: FileBackedOutputStream): SMono[BlobId] =
+    SMono.fromCallable(() => {
+      IOUtils.copy(hashingInputStream, fileBackedOutputStream)
+      Tuples.of(blobIdFactory.from(hashingInputStream.hash.toString), fileBackedOutputStream.asByteSource)
+    })
+      .flatMap((tuple: Tuple2[BlobId, ByteSource]) =>
+        SMono(dumbBlobStore.save(bucketName, tuple.getT1, tuple.getT2))
+          .`then`(SMono.just(tuple.getT1)))
```
Comment (blackheaven) [inline server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala:72]: no `thenReturn`?

File: server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+  override def deleteBucket(bucketName: BucketName): Publisher[Void] = {
+    dumbBlobStore.deleteBucket(bucketName)
+  }
+
+  override def delete(bucketName: BucketName, blobId: BlobId): Publisher[Void] = {
+    Preconditions.checkNotNull(bucketName)
+    Preconditions.checkNotNull(blobId)
+
+    dumbBlobStore.delete(bucketName, blobId)
```
Comment (blackheaven) [inline server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala:97]: deduplicating should not delete, contract tests should be splited.

File: server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala (hunk 1 above)
Comment (chibenwa) [reply server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala:72]: That's not exposed bu SMono :-/

File: server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala (hunk 2 above)
Comment (chibenwa) [reply server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala:97]: > deduplicating should not delete, contract tests should be splited.

See https://github.com/linagora/james-project/issues/3600, what you describe is not achievable for now.

**Comment by chibenwa** [comment]
(only the last fixup, overwize simply rebased on master)

File: server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala [hunk 3]
```diff
@@ -29,6 +29,7 @@ import org.apache.commons.io.IOUtils
 import org.apache.james.blob.api.{BlobId, BlobStore, BucketName, DumbBlobStore}
 import org.reactivestreams.Publisher
 import reactor.core.publisher.Mono
+import reactor.core.scala.publisher.SMono
```
Comment (Arsnael) [inline server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala:32]: Logically you should remove the `import reactor.core.publisher.Mono` above then?

**Comment by blackheaven** [comment]
test this please

File: server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala (hunk 3 above)
Comment (chibenwa) [reply server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala:32]: No it is used for Mono::using

**Comment by chibenwa** [comment]
```
[a73523555384f34713794e4f0847b6994b57c73d] [ERROR]   DistributedTaskManagerTest.listShouldBeEmptyWhenNoTasks 
[a73523555384f34713794e4f0847b6994b57c73d] Expecting empty but was:<[TaskExecutionDetails{taskId=TaskId{value=65d32493-4766-4c94-a825-d5226a90778d}, type=TaskType{typeName=memory-reference-task}, additionalInformation=Optional.empty, status=IN_PROGRESS, submittedDate=2020-07-21T20:51:26.342Z[GMT], submittedNode=Hostname(foo), startedDate=Optional[2020-07-21T20:51:26.446Z[GMT]], ranNode=Optional[Hostname(foo)], completedDate=Optional.empty, canceledDate=Optional.empty, cancelRequestedNode=Optional.empty, failedDate=Optional.empty}]>
```

Isolation issue?

test this please

**Comment by chibenwa** [comment]
```
[a73523555384f34713794e4f0847b6994b57c73d] [ERROR] OpenJDK 64-Bit Server VM warning: INFO: os::commit_memory(0x0000000633c00000, 484442112, 0) failed; error='Not enough space' (errno=12)
[a73523555384f34713794e4f0847b6994b57c73d] [ERROR] OpenJDK 64-Bit Server VM warning: INFO: os::commit_memory(0x0000000633c00000, 484442112, 0) failed; error='Not enough space' (errno=12)#
[a73523555384f34713794e4f0847b6994b57c73d] [INFO] # There is insufficient memory for the Java Runtime Environment to continue.
[a73523555384f34713794e4f0847b6994b57c73d] [INFO] # Native memory allocation (mmap) failed to map 484442112 bytes for committing reserved memory.
[a73523555384f34713794e4f0847b6994b57c73d] [ERROR] # An error report file with more information is saved as:
[a73523555384f34713794e4f0847b6994b57c73d] [INFO] # /james-project/hs_err_pid1391.log
```

test this please

mvn tests :green_apple: can be cunsidered green.

File: server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala (hunk 2 above)
Comment (blackheaven) [reply server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala:98]: From the DOD:

> `DeDuplicationBlobStore::delete` is a noop.

I do not get your comment regarding that

Comment (chibenwa) [reply server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala:98]: I should update the DOD of #3591 to remove this mention then, as doing a noop will, so far, break deletedMessageVault behavior.

Comment (blackheaven) [reply server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala:98]: it should happen in one issue anyway

Comment (blackheaven) [reply server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala:98]: and you may add a disabled test ensuring this behavior

**Comment by chibenwa** [comment]
`RabbitMQDeletedMessageVaultIntegrationTest>DeletedMessageVaultIntegrationTest.vaultEndpointShouldRestoreImapDeletedMailbox:255 » ConditionTimeout`

test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**